exporter/sumologicexporter/sender.go

```diff
@@ -0,0 +1,277 @@
+		droppedRecords []pdata.LogRecord
+		// currentRecords tracks records which are being actually processed
+		currentRecords []pdata.LogRecord
+	)
+
+	for _, record := range s.buffer {
+		data := s.filter.filterOut(record.Attributes())
+		data[logKey] = record.Body().StringVal()
+
+		nextLine, err := json.Marshal(data)
+		if err != nil {
+			droppedRecords = append(droppedRecords, record)
+			errs = append(errs, err)
+			continue
+		}
+
+		sent, appended, err := s.appendAndSend(bytes.NewBuffer(nextLine).String(), LogsPipeline, &body, flds)
+		if err != nil {
+			errs = append(errs, err)
+			if sent {
+				droppedRecords = append(droppedRecords, currentRecords...)
+			}
+
+			if !appended {
+				droppedRecords = append(droppedRecords, record)
+			}
+		}
+
+		// If data was sent, cleanup the currentRecords
+		if sent {
+			currentRecords = currentRecords[:0]
+		}
+
+		// If log has been appended to body, add it to the currentRecords
+		if appended {
+			currentRecords = append(currentRecords, record)
+		}
+	}
+
+	if err := s.send(LogsPipeline, strings.NewReader(body.String()), flds); err != nil {
+		errs = append(errs, err)
+		droppedRecords = append(droppedRecords, currentRecords...)
+	}
+
+	if len(errs) > 0 {
+		return droppedRecords, componenterror.CombineErrors(errs)
+	}
+	return droppedRecords, nil
+}
+
+// appendAndSend appends line to the request body that will be sent and sends
+// the accumulated data if the internal buffer has been filled (with maxBufferSize elements).
+// It returns 2 booleans saying whether the data was sent and whether the provided data
+// was appended and an error if anything failed.
+func (s *sender) appendAndSend(
+	line string,
+	pipeline PipelineType,
+	body *strings.Builder,
+	flds fields,
+) (sent bool, appended bool, _ error) {
+	var errors []error
+	// sent gives information if the data was sent or not
+	sent = false
+	// appended keeps state of appending new log line to the body
+	appended = true
+
+	if body.Len() > 0 && body.Len()+len(line) > s.config.MaxRequestBodySize {
+		sent = true
+		if err := s.send(pipeline, strings.NewReader(body.String()), flds); err != nil {
+			errors = append(errors, err)
+		}
+		body.Reset()
+	}
+
+	if body.Len() > 0 {
+		// Do not add newline if the body is empty
+		if _, err := body.WriteString("\n"); err != nil {
+			errors = append(errors, err)
+			appended = false
+		}
+	}
+
+	if _, err := body.WriteString(line); err != nil && appended {
+		errors = append(errors, err)
+		appended = false
+	}
+
+	if len(errors) > 0 {
+		return sent, appended, componenterror.CombineErrors(errors)
+	}
+	return sent, appended, nil
+}
+
+// cleanBuffer zeroes buffer
+func (s *sender) cleanBuffer() {
+	s.buffer = (s.buffer)[:0]
+}
+
+// batch adds log to the buffer and flushes them if buffer is full to avoid overflow
+// returns list of log records which were not sent successfully
+func (s *sender) batch(log pdata.LogRecord, metadata fields) ([]pdata.LogRecord, error) {
```

Comment:
Is this guaranteed to be called serially? There are quite a few things that could go wrong here if you have concurrent access...

Reply:
Yes, scope is going to be limited to the `pushLogsData`

Reply:
Then it isn't guaranteed to be called serially:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/778b4367f4f3f2a2fd9083b02fe80da062441809/exporter/sumologicexporter/exporter.go#L75-L83

Every time you have a new data point in the pipeline, your `pushLogsData` will be called, and those calls are concurrent.

Reply:
yeah, but sender is going to be created for every `pushLogsData`. Here is more context:

https://github.com/sumo-drosiek/opentelemetry-collector-contrib/blob/8e83f77433dcb5a6554b5538a769970cc66ab8aa/exporter/sumologicexporter/exporter.go#L104-L114

Reply:
Aim of the sender is to be called serially, as concurrency should be handled by `Retry` and `Queue` helpers

Reply:
Alright, makes sense now. Thanks for confirming!